File: src/main/java/com/extendedae_plus/client/render/widgets/button/EAEPIcon.java
package com.extendedae_plus.client.render.widgets.button;

import appeng.client.gui.Icon;
import appeng.client.gui.style.Blitter;
import com.extendedae_plus.ExtendedAEPlus;
import net.minecraft.resources.ResourceLocation;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

public enum EAEPIcon implements IButtonIcon {
    mul2(0, 0),
    div2(16, 0),
    mul3(32, 0),
    div3(48, 0),
    disconnected(64, 0),
    connected(80, 0),
    signalSend(96, 0),
    signalReceive(112, 0),

    mul5(0, 16),
    div5(16, 16),
    patternSingle(32, 16),
    patternMulti(48, 16),
    mergeAdjacency(64, 16),
    mergeNone(80, 16),

    blockingTransparent(0, 32),
    saveCenter(16, 32),
    saveUp(32, 32),
    saveDown(48, 32),

    listWithChildren(0, 48),
    listMulti(16, 48),
    charF(32, 48),
    charL(48, 48),

    ;

    public final int x;
    public final int y;
    public final int width;
    public final int height;

    private static final Logger LOGGER = LoggerFactory.getLogger("[EAEP/Icon]");

    public static final ResourceLocation TEXTURE =
            ExtendedAEPlus.getLocation("textures/gui/icons.png");
    public static final int TEXTURE_WIDTH = 128;
    public static final int TEXTURE_HEIGHT = 128;

    EAEPIcon(int x, int y) {
        this(x, y, 16, 16);
    }

    EAEPIcon(int x, int y, int width, int height) {
        this.x = x;
        this.y = y;
        this.width = width;
        this.height = height;
    }

    @Override
    public Blitter getBlitter() {
        return Blitter.texture(TEXTURE, TEXTURE_WIDTH, TEXTURE_HEIGHT)
                .src(x, y, width, height);
    }

    @Override
    public Icon getAEIcon() {
        LOGGER.warn("Unsupported AEIcon call");
        return Icon.TOOLBAR_BUTTON_BACKGROUND;
    }

    public static IButtonIcon fromAEIcon(Icon aeIcon) {
        return new AEIcon(aeIcon);
    }

    private record AEIcon(Icon aeIcon) implements IButtonIcon {
        @Override
        public Blitter getBlitter() {
            return this.aeIcon.getBlitter();
        }

        @Override
        public Icon getAEIcon() {
            return this.aeIcon;
        }
    }
}

File: src/main/kotlin/com/extendedae_plus/client/render/widgets/button/EAEPIcon.kt
package com.extendedae_plus.client.render.widgets.button

import appeng.client.gui.Icon
import appeng.client.gui.style.Blitter
import com.extendedae_plus.ExtendedAEPlus
import org.slf4j.LoggerFactory

enum class EAEPIcon(val x: Int, val y: Int, val width: Int = 16, val height: Int = 16) : IButtonIcon {
    Mul2(0, 0),
    Div2(16, 0),
    Mul3(32, 0),
    Div3(48, 0),
    Disconnected(64, 0),
    Connected(80, 0),
    SignalSend(96, 0),
    SignalReceive(112, 0),

    Mul5(0, 16),
    Div5(16, 16),
    PatternSingle(32, 16),
    PatternMulti(48, 16),
    MergeAdjacency(64, 16),
    MergeNone(80, 16),

    BlockingTransparent(0, 32),
    SaveCenter(16, 32),
    SaveUp(32, 32),
    SaveDown(48, 32),

    ListWithChildren(0, 48),
    ListMulti(16, 48),
    CharF(32, 48),
    CharL(48, 48),
    ;

    override val blitter: Blitter by lazy {
        Blitter.texture(
            Texture,
            WIDTH,
            HEIGHT
        ).src(x, y, width, height)
    }

    override val aeIcon: Icon
        get() {
            Logger.warn("Unsupported AEIcon call")
            return Icon.TOOLBAR_BUTTON_BACKGROUND
        }

    private data class AEIcon(override val aeIcon: Icon) : IButtonIcon {
        override val blitter: Blitter
            get() = this.aeIcon.blitter
    }

    companion object {
        private val Logger = LoggerFactory.getLogger("EAEP/Icon")

        val Texture = ExtendedAEPlus.getLocation("textures/gui/icons.png")
        const val WIDTH = 128
        const val HEIGHT = 128

        @JvmStatic
        fun fromAEIcon(aeIcon: Icon): IButtonIcon = AEIcon(aeIcon)
    }
}
